### PR TITLE
Interactors

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added interactors

### CODE REVIEW

File: blockchain/txSinger.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+var suite = ed25519.NewEd25519()
+var singleSigner = &singlesig.Ed25519Signer{}
+
+type txSigner struct {
+	keyGen crypto.KeyGenerator
+}
+
+// NewTxSigner contains the primitives used to correctly sign a transaction
```
Comment (bogdan-rosianu) [inline blockchain/txSinger.go:17]: this comment matches better `type txSigner struct`
rephrase to `// NewTxSigner returns a new instance of txSigner`

File: interactors/transaction.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+		return nil, ErrNilTxSigner
+	}
+
+	return &transactionInteractor{
+		Proxy:    proxy,
+		TxSigner: txSigner,
+	}, nil
+}
+
+func (ti *transactionInteractor) SetMillisecondsBetweenBunches(millisecondsBetweenBunches uint32) {
```
Comment (bogdan-rosianu) [inline interactors/transaction.go:56]: add a comment to exported functions.
Also, please set a default value for `ti.millisecondsBetweenBunches` on the constructor. if this setter is not called, then there will be no sleeping period

File: interactors/transaction.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+		return ArgCreateTransaction{}, err
+	}
+
+	copyArg.Signature = hex.EncodeToString(signature)
+
+	return copyArg, nil
+}
+
+func (ti *transactionInteractor) createUnsignedMessage(arg ArgCreateTransaction) ([]byte, error) {
+	tx := ti.CreateTransaction(arg)
```
Comment (bogdan-rosianu) [inline interactors/transaction.go:126]: add a `arg.Signature = ""` before just to be sure

File: interactors/transaction.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+func (ti *transactionInteractor) SendTransactionsAsBunch(bunchSize int) ([]string, error) {
+	if bunchSize <= 0 {
+		return nil, ErrInvalidValue
+	}
+
+	millisecondsBetweenBunches := atomic.LoadUint32(&ti.millisecondsBetweenBunches)
+
+	transactions := ti.PopAccumulatedTransactions()
+	allHashes := make([]string, 0)
+	for bunchIndex := 0; len(transactions) > 0; bunchIndex++ {
```
Comment (bogdan-rosianu) [inline interactors/transaction.go:140]: interesting approach ! :+1:

File: interactors/transaction_test.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+	sk, err := hex.DecodeString("6ae10fed53a84029e53e35afdbe083688eea0917a09a9431951dd42fd4da14c40d248169f4dd7c90537f05be1c49772ddbf8f7948b507ed17fb23284cf218b7d")
+
+	assert.Nil(t, err)
+
+	ti, err := NewTransactionInteractor(proxy, txSigner)
+	assert.Nil(t, err)
+	assert.NotNil(t, ti)
+
+	value := big.NewInt(0)
```
Comment (bogdan-rosianu) [inline interactors/transaction_test.go:29]: could have been one liner

File: interactors/transaction_test.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+			return msgs, nil
+		},
+	}
+
+	var signer TxSigner = &mock.TxSignerStub{}
+
+	ti, err := NewTransactionInteractor(proxy, signer)
+	assert.Nil(t, err, "Error on transaction interactor constructor")
+
+	value := big.NewInt(0)
```
Comment (bogdan-rosianu) [inline interactors/transaction_test.go:73]: one liner

File: interactors/transaction_test.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+		0,
+	}
+	tx := ti.CreateTransaction(args)
+	ti.AddTransaction(tx)
+
+	msg, err := ti.SendTransactionsAsBunch(1)
+	assert.Nil(t, err, "Error on sending bunches")
+	fmt.Println("Message: ", msg)
+	assert.NotNil(t, msg)
+
```
Comment (bogdan-rosianu) [inline interactors/transaction_test.go:95]: remove an empty line + maybe remove prints

File: interactors/transaction_test.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+			return msgs, nil
+		},
+	}
+
+	var signer TxSigner = &mock.TxSignerStub{}
+
+	ti, err := NewTransactionInteractor(proxy, signer)
+	assert.Nil(t, err, "Error on transaction interactor constructor")
+
+	value := big.NewInt(0)
```
Comment (bogdan-rosianu) [inline interactors/transaction_test.go:116]: one liner

File: interactors/transaction_test.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	assert.Equal(t, 1, sendCalled)
+	assert.Nil(t, err)
+
+	sendCalled = 0
+	numTxs = 101
+	for i := 0; i < numTxs; i++ {
+		ti.AddTransaction(&data.Transaction{})
+	}
+	hashes, err = ti.SendTransactionsAsBunch(2)
+	assert.Equal(t, 101, len(hashes))
```
Comment (bogdan-rosianu) [inline interactors/transaction_test.go:198]: `numTxs` instead of `101`

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: interactors/transaction.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+	pkBytes, err := ti.TxSigner.GeneratePkBytes(skBytes)
+	if err != nil {
+		return ArgCreateTransaction{}, err
+	}
+
+	copyArg := arg
+	copyArg.Signature = ""
+	copyArg.SndAddr = core.AddressPublicKeyConverter.Encode(pkBytes)
+
+	unsignedMessaged, err := ti.createUnsignedMessage(copyArg)
```
Comment (SebastianMarian) [inline interactors/transaction.go:110]: unsignedMessage

File: interactors/transaction_test.go (hunk 7 above)
Comment (SebastianMarian) [inline interactors/transaction_test.go:95]: Remove empty line

File: interactors/transaction_test.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+		0,
+	}
+
+	args, err = ti.ApplySignatureAndSender(sk, args)
+	tx := ti.CreateTransaction(args)
+
+	assert.Equal(t, "erd1p5jgz605m47fq5mlqklpcjth9hdl3au53dg8a5tlkgegfnep3d7stdk09x", tx.SndAddr)
+	assert.Equal(t, "80e1b5476c5ea9567614d9c364e1a7380b7990b53e7b6fd8431bf8536d174c8b3e73cc354b783a03e5ae0a53b128504a6bcf32c3b9bbc06f284afe1fac179e0d",
+		tx.Signature)
+
```
Comment (SebastianMarian) [inline interactors/transaction_test.go:52]: Remove empty line

File: blockchain/txSinger.go (hunk 1 above)
Comment (iulianpascalau) [reply blockchain/txSinger.go:17]: done

File: interactors/transaction.go (hunk 3 above)
Comment (iulianpascalau) [reply interactors/transaction.go:126]: done

File: interactors/transaction_test.go (hunk 5 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:29]: old code, refactored

File: interactors/transaction_test.go (hunk 6 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:73]: done

File: interactors/transaction_test.go (hunk 7 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:95]: done

File: interactors/transaction_test.go (hunk 8 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:116]: done

File: interactors/transaction_test.go (hunk 9 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:198]: done

File: interactors/transaction.go (hunk 10 above)
Comment (iulianpascalau) [reply interactors/transaction.go:110]: fixed

File: interactors/transaction_test.go (hunk 7 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:95]: fixed

File: interactors/transaction_test.go (hunk 11 above)
Comment (iulianpascalau) [reply interactors/transaction_test.go:52]: fixed

**Review by bogdan-rosianu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**